### PR TITLE
Don't adjust offsets in apply_ygain_zoom for TraceViewer

### PR DESCRIPTION
**Pull request by jpgill86**
The objective of the code removed by this commit was to keep the medians of the signals fixed in place as ygain was zoomed in or out under `by_channel` or `same_for_all` scale modes. However, because the median is only estimated from a randomly selected subset of samples and also changes from chunk to chunk, this makes ygain zoom in these modes a non-reversible operation, i.e., zooming in and then zooming out does not return you to the same configuration. Instead, the offsets of the signals can drift away from their original locations and the signals can even move out of view, like this:

ygain_zoom_with_offsets

Notice that the channel labels, which are anchored to the offsets, move around during zooming. After zooming out again the channel spacing is not the same as it was before, and the bottom gray trace has completely migrated off the screen!

It's because of this behavior that I have strictly avoided ever using the ygain zoom if I could help it!

This commit changes the behavior of ygain zoom in `by_channel` and `same_for_all` scale modes by making it leave the offsets unchanged. Consequently, signals are scaled around 0 (baseline) instead of their medians, and the zoom is perfectly reversible:

ygain_zoom_without_offsets

Notice here the labels don't move, and everything returns to its original place in the end.

A possibly undesired consequence of this change is that signals with a large, meaningless non-zero baseline, such as a channel with a spurious DC offset, will not expand and contract around its median anymore, but instead will move around the screen. @samuelgarcia, how much are you worried about this? I could instead instantiate a new scale mode that has the behavior I desire, but that seems like a lot more work. 🤣

### CODE REVIEW

**Comment by jpgill86** [comment]
P.S.: Sorry that I'm giving you so much to review lately! I know you probably don't have a lot of time for this!

**Comment by samuelgarcia** [comment]
This is so perfect, I want to fix this since a long long time. But never find time to do it.
Give some days to check the solution.

**Comment by samuelgarcia** [comment]
I really don't anderstand the solution :  you remove the offset compensation when gain moving and now the offset is stable ????

Did you check this with a signal that is not centered on 0 ?

**Comment by jpgill86** [comment]
TL;DR: I should probably just change this into a new option that is disabled by default.

> I really don't anderstand the solution [...] Did you check this with a signal that is not centered on 0 ?

You are puzzled because I've changed the code to no longer do what perhaps you consider to be a critical thing: Keep the signal centered as the zoom changes. By removing offset compensation, signals close to zero will not move much, but signals far from zero will, and you are worried about this.

In some cases this new behavior is desired. For example, the bottom gray trace in my animated GIFs is a force measurement. In my setup, force is always nonnegative, so its median is far from zero, and yet removing offset compensation is, in this case, what I want. The reason is because deviation from zero is meaningful, so I want to keep the zero-baseline visible on screen at all times and have the force trace scale upwards from there.

Contrast this to an extracellular voltage signal that has roughly equal sizes of positive and negative deflections. Here the median should ideally be zero, and removal of offset compensation will not matter. However, in reality sometimes such signals have a DC offset such that the signals are centered on some value other than zero. These signals will appear to drift up or down the screen as the user changes the ygain zoom because the zero stays in one place as the signal is rescaled. I am finding that for my data sets the non-ideal DC offset is very small compared to size of the signal, so this is not a problem.

However, if you have signals that are centered on values far from zero (like my force measurements), and that deviation from zero is not meaningful (_unlike_ my force measurements), such as a very large DC offset on an extracellular signal, you really need offset compensation to keep your signal centered as the zoom changes.

Perhaps this last case is more common than I anticipate, and removing offset compensation completely is a very bad idea. If you think that's the case @samuelgarcia, I could make offset compensation into a new option called "`ygain_zoom_around_median`" (alternative names welcome), which could be enabled by default. Disabling this new option would disable offset compensation and would satisfy my needs without breaking ephyviewer for others.

What do you think?

**Comment by samuelgarcia** [comment]
So I think we must have this option.
EEG data for instance have always crazy offest and we want to have then centered.

**ygain_zoom_around_median** or **offset_compensate_when_zoom**, as you want. I don't shorter idea.

An idea for would be to no use the median which computed with the begining but the first displayed sample for each channels. So no computation at all just refresh this value and time is changing and the offset compensation would be better when the median is moving.

What do you think ?

**Comment by jpgill86** [comment]
This is funny, I just realized there are two mistakes in the original offset compensation code that are the main reason the original rescaling around the median worked so poorly!

We want this to be the result of zooming:

offset_new + center * gain_new = offset_old + center * gain_old

Here "center" is whatever value the zoom is centered on. This was originally the median. I changed it to zero (and hence the term dropped out, and offsets became unchanging). You are now suggesting the first sample value, which I think could work.

But still there is an error in the code that does this calculation! We compute

gain_new = gain_old * factor_ratio

Putting this into the first equation and solving for offset_new gives

offset_new = offset_old + center * gain_old * (1 - factor_ratio)

OK, that looks like this original code, right?

https://github.com/NeuralEnsemble/ephyviewer/blob/065ab6224e6503d2f67a8997f6a372b2e18ce94b/ephyviewer/traceviewer.py#L187-L190

Wrong! Because `self.gains` is used in line 190 after being changed in line 189, we're actually calculating

offset_new = offset_old + center * **gain_new** * (1 - factor_ratio)

with gain_new instead of gain_old!

I fixed this on my machine and the lines are much more stationary, though not perfectly so.

The last issue is this code:

https://github.com/NeuralEnsemble/ephyviewer/blob/065ab6224e6503d2f67a8997f6a372b2e18ce94b/ephyviewer/traceviewer.py#L187-L188

That should be `hasattr(self, 'signals_med')` instead. As it was, `estimate_median_mad` was getting re-executed with every call to `apply_ygain_zoom`, even if the chunk has not changed. Not only is this extra work the program shouldn't be doing, but it also means that the "center" is shifting around a little bit all the time. This causes the remainder of the drift in signal position.

What we need is for the median to be estimated just once per chunk, and have the zoom expand or contract around that "center" until the user pans to a new part of the signal and requests a new chunk. Does that sound right to you, @samuelgarcia ? I'm going to work on implementing it to see how well I like it. The signals will still be capable of drifting if the user zooms in and out a lot while moving through several chunks, since the median will change, but in some cases we may want that. I'll try it for my data and see if I like it.

In addition to fixing these two things, we could also have an option for controlling whether "center" is the estimated median, the first sample, or zero.

**Comment by samuelgarcia** [comment]
Oh my god.
I was so stupid.

I think just now we can:
  * 'self.signals_med' > 'signals_med'
  * swap line gain and offset

I think I have the same in pyacq!

If you still think we need an option to remove the comasention add it.

You are the new president of ephyviewer for one year at least.
And you come and visit Lyon, you will have a glass of good wine :)

**Comment by jpgill86** [comment]
> I think just now we can:
> - 'self.signals_med' > 'signals_med'
> - swap line gain and offset

Yes, that is the simplest thing, and I will do that. In my last post I proposed re-estimating the median once per chunk, so that if you stepped to a different time it would redo the median calculation once before zooming. But after looking at the code I see it's not quite as simple as I thought. And I think this simple solution you recommend is good enough. The medians will be estimated exactly once unless auto-scale is clicked or scale mode is changed.

This also seems to be good enough for my data sets for now, even without changing the "center" to 0. Maybe I'll change my mind later, but for now I'll just fix these bugs. 😊

As my first act as the American president of ephyviewer, I will withdraw us from the NeuralEnsemble.... just kidding!
